server/src/main/java/io/crate/execution/engine/aggregation/impl/ArbitraryAggregation.java

```diff
@@ -108,4 +125,97 @@ public Object reduce(RamAccounting ramAccounting, Object state1, Object state2)
     public Object terminatePartial(RamAccounting ramAccounting, Object state) {
         return state;
     }
+
+    @Nullable
+    @Override
+    public DocValueAggregator<?> getDocValueAggregator(List<DataType<?>> argumentTypes,
+                                                       List<MappedFieldType> fieldTypes) {
+        var dataType = argumentTypes.get(0);
+        switch (dataType.id()) {
+            case ByteType.ID:
+            case ShortType.ID:
+            case IntegerType.ID:
+            case LongType.ID:
+            case TimestampType.ID_WITH_TZ:
+            case TimestampType.ID_WITHOUT_TZ:
+                return new ArbitraryNumericDocValueAggregator(
+                    fieldTypes.get(0).name(),
+                    dataType,
+                    (values, state) -> {
+                        if (!state.hasValue()) {
+                            state.setValue(values.nextValue());
+                        }
+                    }
+                );
+            case FloatType.ID:
+                return new ArbitraryNumericDocValueAggregator(
+                    fieldTypes.get(0).name(),
+                    dataType,
+                    (values, state) -> {
+                        if (!state.hasValue()) {
+                            var value = NumericUtils.sortableIntToFloat((int) values.nextValue());
+                            state.setValue(value);
+                        }
+                    }
+                );
+            case DoubleType.ID:
+                return new ArbitraryNumericDocValueAggregator(
+                    fieldTypes.get(0).name(),
+                    dataType,
+                    (values, state) -> {
+                        if (!state.hasValue()) {
+                            var value = NumericUtils.sortableLongToDouble(values.nextValue());
+                            state.setValue(value);
+                        }
+                    }
+                );
+            case IpType.ID:
+            case StringType.ID:
+                return new BinaryDocValueAggregator<>(
+                    fieldTypes.get(0).name(),
+                    MutableObject::new,
```

Comment:
I think we could use the value directly without the wrapper in this case. There is no primitive type involved, so it doesn't avoid boxing. Or am I missing something?

Reply:
But I still need to have the state represented by some kind of a reference holder, otherwise, if it would be e.g. just an object value it won't be possible to set/point the state to the extracted doc value.

Reply:
Ah right, the `apply` of the `DocValueAggregator` only works with side effects.